x/upgrade/keeper/keeper.go

```diff
@@ -69,13 +73,48 @@ func (k Keeper) ScheduleUpgrade(ctx sdk.Context, plan types.Plan) error {
 		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "upgrade with name %s has already been completed", plan.Name)
 	}
 
-	bz := k.cdc.MustMarshalBinaryBare(&plan)
 	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshalBinaryBare(&plan)
 	store.Set(types.PlanKey(), bz)
 
+	if plan.UpgradedClientState != nil {
+		clientState, err := clienttypes.UnpackClientState(plan.UpgradedClientState)
+		if err != nil {
+			return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "could not unpack clientstate: %v", err)
+		}
+		return k.SetUpgradedClient(ctx, clientState)
+	}
+	// delete upgraded client key to remove any upgraded client set by outdated plan
+	store.Delete(types.UpgradedClientKey())
+	return nil
+}
+
+// SetUpgradedClient sets the expected upgraded client for the next version of this chain
+func (k Keeper) SetUpgradedClient(ctx sdk.Context, cs ibcexported.ClientState) error {
+	// zero out any custom fields before setting
+	cs = cs.ZeroCustomFields()
+	bz, err := clienttypes.MarshalClientState(k.cdc, cs)
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "could not marshal clientstate: %v", err)
```

Comment:
ditto

Reply:
Not really possible for this to ever fail (without an incorrect `ZeroCustomFields` implementation), just a sanity check